### PR TITLE
Support Ip-adapter_face_id

### PR DESCRIPTION
**Pull request by sdbds**
To make it easier to use, i added scikit-image and converted insightface's lib to local file to be used without additional installation.
I would like to discuss this because I have some questions about the design.
I've talked to the author of Ip-adapter, do I need to add the Lora part? 
Because this part is not included in controlnet. 
In addition, because there are some hack and hook parts, this part of the webui code I dont know, I hope @lllyasviel can modify it.

### CODE REVIEW

**Comment by xiaohu2015** [comment]
yes,  a standard IP-Adapter + a standard LoRA

File: annotator/insightface/face_analysis.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+# @Function      : 
+
+
+from __future__ import division
+
+import glob
+import os.path as osp
+
+import numpy as np
+import onnxruntime
```
Comment (huchenlei) [inline annotator/insightface/face_analysis.py:14]: Add onnx/onnxruntime to `requirements.txt`.

File: scripts/processor.py [hunk 2]
```diff
@@ -639,6 +639,15 @@ def unload_anime_face_segment():
     if model_anime_face_segment is not None:
         model_anime_face_segment.unload_model()
 
+def face_id(img, res=512, plus=False, **kwargs):
+    img = HWC3(img)
+    global model_face_id
+    if model_face_id is None:
+        from annotator.insightface import FaceidEmbedsEstimator
+        model_face_id = FaceidEmbedsEstimator()
```
Comment (huchenlei) [inline scripts/processor.py:647]: `model_face_id` is not defined in global scope.

File: annotator/insightface/model_zoo/inswapper.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
+import time
+import numpy as np
+import onnxruntime
+import cv2
+import onnx
+from onnx import numpy_helper
+from ..face_align import face_align
```
Comment (huchenlei) [inline annotator/insightface/model_zoo/inswapper.py:7]: `face_align` not defined in `face_align.py`.

File: annotator/insightface/__init__.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+import torch
+from PIL import Image
+
+import sys
+
+import numpy as np
+from modules import devices
+from .face_analysis import FaceAnalysis
+import face_align
```
Comment (huchenlei) [inline annotator/insightface/__init__.py:12]: To `from .face_align import norm_crop`

File: scripts/controlmodel_ipadapter_face_id.py [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+
+import torch
+from PIL import Image
+from safetensors import safe_open
+
+import sys
+
+import numpy as np
+from modules import devices
+from .controlmodel_ipadapter import PerceiverAttention, FeedForward
```
Comment (huchenlei) [inline scripts/controlmodel_ipadapter_face_id.py:13]: To `from scripts.controlmodel_ipadapter import PerceiverAttention, FeedForward`

File: annotator/insightface/face_analysis.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                print('model ignore:', onnx_file, model.taskname)
+                del model
+            elif model.taskname not in self.models and (allowed_modules is None or model.taskname in allowed_modules):
+                print('find model:', onnx_file, model.taskname, model.input_shape, model.input_mean, model.input_std)
+                self.models[model.taskname] = model
+            else:
+                print('duplicated model task type, ignore:', onnx_file, model.taskname)
+                del model
+        assert 'detection' in self.models
+        self.det_model = self.models['detection']
```
Comment (huchenlei) [inline annotator/insightface/face_analysis.py:46]: I get following error
```
download_path: D:\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\downloads\insightface\models\buffalo_l
Downloading D:\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\downloads\insightface\models\buffalo_l.zip from https://github.com/deepinsight/insightface/releases/download/v0.7/buffalo_l.zip...
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 281857/281857 [00:56<00:00, 4946.30KB/s]
Applied providers: ['CUDAExecutionProvider', 'CPUExecutionProvider'], with options: {'CUDAExecutionProvider': {'do_copy_in_default_stream': '1', 'cudnn_conv_algo_search': 'EXHAUSTIVE', 'device_id': '0', 
'gpu_external_alloc': '0', 'enable_cuda_graph': '0', 'gpu_mem_limit': '18446744073709551615', 'gpu_external_free': '0', 'gpu_external_empty_cache': '0', 'arena_extend_strategy': 'kNextPowerOfTwo', 'cudnn_conv_use_max_workspace': '1', 'cudnn_conv1d_pad_to_nc1d': '0', 'tunable_op_enable': '0', 'tunable_op_tuning_enable': '0', 'enable_skip_layer_norm_strict_mode': '0'}, 'CPUExecutionProvider': {}}
model not recognized: D:\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\downloads\insightface\models\buffalo_l\1k3d68.onnx
Applied providers: ['CUDAExecutionProvider', 'CPUExecutionProvider'], with options: {'CUDAExecutionProvider': {'do_copy_in_default_stream': '1', 'cudnn_conv_algo_search': 'EXHAUSTIVE', 'device_id': '0', 
'gpu_external_alloc': '0', 'enable_cuda_graph': '0', 'gpu_mem_limit': '18446744073709551615', 'gpu_external_free': '0', 'gpu_external_empty_cache': '0', 'arena_extend_strategy': 'kNextPowerOfTwo', 'cudnn_conv_use_max_workspace': '1', 'cudnn_conv1d_pad_to_nc1d': '0', 'tunable_op_enable': '0', 'tunable_op_tuning_enable': '0', 'enable_skip_layer_norm_strict_mode': '0'}, 'CPUExecutionProvider': {}}
model not recognized: D:\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\downloads\insightface\models\buffalo_l\2d106det.onnx
Applied providers: ['CUDAExecutionProvider', 'CPUExecutionProvider'], with options: {'CUDAExecutionProvider': {'do_copy_in_default_stream': '1', 'cudnn_conv_algo_search': 'EXHAUSTIVE', 'device_id': '0', 
'gpu_external_alloc': '0', 'enable_cuda_graph': '0', 'gpu_mem_limit': '18446744073709551615', 'gpu_external_free': '0', 'gpu_external_empty_cache': '0', 'arena_extend_strategy': 'kNextPowerOfTwo', 'cudnn_conv_use_max_workspace': '1', 'cudnn_conv1d_pad_to_nc1d': '0', 'tunable_op_enable': '0', 'tunable_op_tuning_enable': '0', 'enable_skip_layer_norm_strict_mode': '0'}, 'CPUExecutionProvider': {}}
model not recognized: D:\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\downloads\insightface\models\buffalo_l\det_10g.onnx
Applied providers: ['CUDAExecutionProvider', 'CPUExecutionProvider'], with options: {'CUDAExecutionProvider': {'do_copy_in_default_stream': '1', 'cudnn_conv_algo_search': 'EXHAUSTIVE', 'device_id': '0', 
'gpu_external_alloc': '0', 'enable_cuda_graph': '0', 'gpu_mem_limit': '18446744073709551615', 'gpu_external_free': '0', 'gpu_external_empty_cache': '0', 'arena_extend_strategy': 'kNextPowerOfTwo', 'cudnn_conv_use_max_workspace': '1', 'cudnn_conv1d_pad_to_nc1d': '0', 'tunable_op_enable': '0', 'tunable_op_tuning_enable': '0', 'enable_skip_layer_norm_strict_mode': '0'}, 'CPUExecutionProvider': {}}
model not recognized: D:\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\downloads\insightface\models\buffalo_l\genderage.onnx
Applied providers: ['CUDAExecutionProvider', 'CPUExecutionProvider'], with options: {'CUDAExecutionProvider': {'do_copy_in_default_stream': '1', 'cudnn_conv_algo_search': 'EXHAUSTIVE', 'device_id': '0', 
'gpu_external_alloc': '0', 'enable_cuda_graph': '0', 'gpu_mem_limit': '18446744073709551615', 'gpu_external_free': '0', 'gpu_external_empty_cache': '0', 'arena_extend_strategy': 'kNextPowerOfTwo', 'cudnn_conv_use_max_workspace': '1', 'cudnn_conv1d_pad_to_nc1d': '0', 'tunable_op_enable': '0', 'tunable_op_tuning_enable': '0', 'enable_skip_layer_norm_strict_mode': '0'}, 'CPUExecutionProvider': {}}
model not recognized: D:\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\downloads\insightface\models\buffalo_l\w600k_r50.onnx
Traceback (most recent call last):
  File "d:\stable-diffusion-webui\venv\lib\site-packages\gradio\routes.py", line 488, in run_predict
    output = await app.get_blocks().process_api(
  File "d:\stable-diffusion-webui\venv\lib\site-packages\gradio\blocks.py", line 1431, in process_api
    result = await self.call_function(
  File "d:\stable-diffusion-webui\venv\lib\site-packages\gradio\blocks.py", line 1103, in call_function
    prediction = await anyio.to_thread.run_sync(
  File "d:\stable-diffusion-webui\venv\lib\site-packages\anyio\to_thread.py", line 33, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "d:\stable-diffusion-webui\venv\lib\site-packages\anyio\_backends\_asyncio.py", line 877, in run_sync_in_worker_thread
    return await future
  File "d:\stable-diffusion-webui\venv\lib\site-packages\anyio\_backends\_asyncio.py", line 807, in run
    result = context.run(func, *args)
  File "d:\stable-diffusion-webui\venv\lib\site-packages\gradio\utils.py", line 707, in wrapper
    response = f(*args, **kwargs)
  File "D:\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\controlnet_ui\controlnet_ui_group.py", line 694, in run_annotator
    result, is_image = preprocessor(
  File "D:\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\utils.py", line 75, in decorated_func
    return cached_func(*args, **kwargs)
  File "D:\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\utils.py", line 63, in cached_func
    return func(*args, **kwargs)
  File "D:\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\global_state.py", line 37, in unified_preprocessor
    return preprocessor_modules[preprocessor_name](*args, **kwargs)
  File "D:\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\processor.py", line 651, in face_id
    model_face_id = FaceidEmbedsEstimator()
  File "D:\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\insightface\__init__.py", line 18, in __init__
    self.model = FaceAnalysis(name="buffalo_l", providers=['CUDAExecutionProvider', 'CPUExecutionProvider'])
  File "D:\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\insightface\face_analysis.py", line 45, in __init__
    assert 'detection' in self.models
AssertionError
```

**Review by huchenlei — CHANGES_REQUESTED** [review]

**Comment by huchenlei** [comment]
Continued in #2434.

**Pull request closed without merging**